6676 port generative networks spade

marksgraham opened the pull request:
Towards #6676  .

### Description

This adds SPADE-enabled autoencoder and diffusion_model_unet architectures. They are new implementations for each network, rather than options in the existing network, because @virginiafdez and I felt that adding additional options to the existing networks to enable spade compatibility significantly reduced the readability of them for users who were not interested in SPADE functionality.

These are the last networks to be ported over.

### Types of changes
<!--- Put an `x` in all the boxes that apply, and remove the not applicable items -->
- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

ericspod commented on monai/networks/blocks/spade_norm.py:
```diff
@@ -0,0 +1,97 @@
+            norm_params = {}
+        if len(norm_params) != 0:
+            norm = (norm, norm_params)
+        self.param_free_norm = ADN(
```
We could just use `get_norm_layer` instead of `ADN`.

marksgraham replied:
added to refactor issue

ericspod commented on monai/networks/blocks/spade_norm.py:
```diff
@@ -0,0 +1,97 @@
+
+class SPADE(nn.Module):
+    """
+    SPADE normalisation block based on the 2019 paper by Park et al. (doi: https://doi.org/10.48550/arXiv.1903.07291)
```
Could we add a little more description about what the block is meant for and what it does?

ericspod commented on monai/networks/blocks/spade_norm.py:
```diff
@@ -0,0 +1,97 @@
+        spatial_dims: int = 2,
+        hidden_channels: int = 64,
+        norm: str | tuple = "INSTANCE",
+        norm_params: dict | None = None,
```
As with the `norm` argument [here](https://github.com/marksgraham/MONAI/blob/035d2d2764c8db2836f5d2af4418a4d19fd954d6/monai/networks/blocks/convolutions.py#L107C10-L107C10) this separate parameters shouldn't be needed I don't think. `norm` is either just the name or a tuple containing the name and then whatever `norm_params` would have so I don't think the separate argument is needed.

marksgraham replied:
added to refactor list

ericspod commented on monai/networks/blocks/spade_norm.py:
```diff
@@ -0,0 +1,97 @@
+            out_channels=hidden_channels,
+            kernel_size=kernel_size,
+            norm=None,
+            padding=kernel_size // 2,
```
I don't think we need this argument since `same_padding` will be used if we provide nothing. This should give the same results for odd-sized kernels, for even ones the behaviour would be different but I don't think we'd do that?

ericspod reviewed (approved):
Looks good for merging into the branch. We can just keep track of the things I mentioned.

KumoLiu commented on docs/source/networks.rst:
```diff
@@ -248,6 +248,7 @@ Blocks
 .. autoclass:: monai.apps.reconstruction.networks.blocks.varnetblock.VarNetBlock
    :members:
 
+
```
Add this line by accident?

KumoLiu commented on monai/networks/blocks/spade_norm.py:
```diff
@@ -0,0 +1,97 @@
+        """
+        Args:
+            x: input tensor
+            segmap: input segmentation map (bxcx[spatial-dimensions]) where c is the number of semantic channels.
```
```suggestion
            segmap: input segmentation map (B, C, [spatial-dimensions]) where C is the number of semantic channels.
```

KumoLiu commented on monai/networks/blocks/spade_norm.py:
```diff
@@ -0,0 +1,97 @@
+            kernel_size=kernel_size,
+            norm=None,
+            padding=kernel_size // 2,
+            act="LEAKYRELU",
```
Is it hardcoded because of the original implementation? Maybe we should make it a default? What do you think?

ericspod replied:
This could be a parameter yes.

KumoLiu commented on monai/networks/blocks/spade_norm.py:
```diff
@@ -0,0 +1,97 @@
+    def forward(self, x: torch.Tensor, segmap: torch.Tensor) -> torch.Tensor:
+        """
+        Args:
+            x: input tensor
```
```suggestion
            x: input tensor with shape (B, C, [spatial-dimensions]) where C is the number of semantic channels.
```

KumoLiu commented on monai/networks/nets/spade_autoencoderkl.py:
```diff
@@ -0,0 +1,473 @@
+            with_nonlocal_attn=with_encoder_nonlocal_attn,
+            use_flash_attention=use_flash_attention,
+        )
+        self.decoder = SPADEDecoder(
```
If I understand correctly, the only difference is this decoder module. Or if I misunderstand, feel free to correct.
I personally feel we could combine this network and `AutoencoderKL` into one, looks there are many duplicated code. I think we can simply add a check here, my feeling is that dividing them into two networks will require users to distinguish between them on their own, which is not straightforward.

cc @ericspod, @Nic-Ma for more discussion here.

ericspod replied:
Yes that'll be something to look into when refactoring the feature branch.

marksgraham replied:
I agree, it isn't all that different. Its the diffusion_model_unet where the real differences between the SPADE/non-SPADE networks are. I have added merging the normal and SPADE autoencoders to the list of items to be done in the refactor

KumoLiu commented on monai/networks/nets/spade_diffusion_model_unet.py:
```diff
@@ -0,0 +1,909 @@
+    Unet network with timestep embedding and attention mechanisms for conditioning based on
+    Rombach et al. "High-Resolution Image Synthesis with Latent Diffusion Models" https://arxiv.org/abs/2112.10752
+    and Pinaya et al. "Brain Imaging Generation with Latent Diffusion Models" https://arxiv.org/abs/2209.07162
+    Enables SPADE normalisation for semantic conditioning (Park et. al (2019): https://github.com/NVlabs/SPADE)
```
Do we have some corresponding tutorial about these networks? If yes, would be better to mention the link here which may help users get started.

marksgraham replied:
Sure, added a link

KumoLiu reviewed (commented):
Thanks for the PR!
I want to confirmed that the refactor will also be done in this branch, as previously discussed, right?

ericspod commented:
> Thanks for the PR and hard work!
> I want to confirmed that the refactor will also be done in this branch, as previously discussed, right?

We definitely will be refactoring a lot of things in this feature branch before merging into dev. There's a lot of structural overlap between networks and redundant block definitions we can consolidate. We'll also do a pass over documentation to make sure it's consistent and complete.

marksgraham commented:
Hi @KumoLiu 
Thanks for the feedback - I've addressed the minor bits and added the others to the refactoring issue, [https://github.com/Project-MONAI/MONAI/issues/7227](https://github.com/Project-MONAI/MONAI/issues/7227). 

Just a few more components to add (noise schedulers for diffusion models, some Inferers and Engines) and then we can begin the refactor

KumoLiu commented:
/build
